README.md
# NAME

Badge::Depot - A framework for badges

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/Csson/p5-Badge-Depot"><img src="https://api.travis-ci.org/Csson/p5-Badge-Depot.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Badge-Depot-0.0104"><img src="https://badgedepot.code301.com/badge/kwalitee/Badge-Depot/0.0104" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Badge-Depot%200.0104"><img src="https://badgedepot.code301.com/badge/cpantesters/Badge-Depot/0.0104" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-100.0%-brightgreen.svg" alt="coverage 100.0%" />
    </p>
</div>

# VERSION

Version 0.0104, released 2016-02-19.

# SYNOPSIS

    # Define a badge class
    package Badge::Depot::Plugin::Example;

    use Moose;
    with 'Badge::Depot';

    has user => (
        is => 'ro',
        isa => 'Str',
        required => 1,
    );

    sub BUILD {
        my $self = shift;
        $self->link_url(sprintf 'https://example.com/users/%s', $self->user);
        $self->image_url(sprintf 'https://example.com/users/%s.svg', $self->user);
        $self->image_alt('Example text');
    }

    # Somewhere else
    my $badge = Badge::Depot::Plugin::Example->new(user => 'my_username');

    print $badge->to_html;
    # prints '<a href="https://example.com/users/my_username"><img src="https://example.com/users/my_username.svg" alt="Example text" /></a>'

# DESCRIPTION

`Badge::Depot` is a framework for documentation badges. Using badges in your documentation can give
end users of your distribution dynamic information without you having to update the documentation.

You only need use this distribution directly if you want to create a new badge in the `Badge::Depot::Plugin` namespace. See [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot) for
a list of existing badges.

For use together with [Pod::Weaver](https://metacpan.org/pod/Pod::Weaver), see [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges).

# OVERVIEW

`Badge::Depot` is a [Moose](https://metacpan.org/pod/Moose) role that adds a few attributes and methods.

# ATTRIBUTES

These attributes are expected to be set when the badge class returns from `new`. See [synopsis](#synopsis).

## image\_url

Required. [Uri](https://metacpan.org/pod/Types::URI).

The url to the actual badge. The src attribute for the img tag when rendered to html.

## image\_alt

Optional. [Str](https://metacpan.org/pod/Types::Standard).

The alternative text of the badge. The alt attribute for the img tag when rendered to html. No alternative text is created if this isn't set.

## link\_url

Optional (but recommended). [Uri](https://metacpan.org/pod/Types::URI).

The url to link the badge to. The href attribute for the a tag when rendered to html. No link is created if this isn't set.

## zilla

This is not a user setable attribute, but rather can be used by plugins that might need to look at distribution meta data. This is only usable when using [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges) together with [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla).

# METHODS

These methods are used when rendering the badge, and are not useful inside badge classes.

## to\_html

Returns a string with the badge rendered as html.

## to\_markdown

Returns a string with the badge rendered as markdown.

# SEE ALSO

- [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot) - List of available badges
- [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges) - Weave the badges
- [WWW::StatusBadge](https://metacpan.org/pod/WWW::StatusBadge) - An alternative badge framework

# SOURCE

[https://github.com/Csson/p5-Badge-Depot](https://github.com/Csson/p5-Badge-Depot)

# HOMEPAGE

[https://metacpan.org/release/Badge-Depot](https://metacpan.org/release/Badge-Depot)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Badge-Depot

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0104 2016-02-19T22:06:00Z UTC
 [New Features]
 - New attribute 'zilla' that exposes the Dist::Zilla object if
   Badge::Depot is used via Pod::Weaver::Section::Badges during a
   Dist::Zilla build.

 [Stats & Dependency Changes]
 - 3 files changed, 22 insertions(+), 12 deletions(-)
 - (run req) ~ MooseX::AttributeShortcuts (any) --> 0.028
 - (run req) ~ Types::Standard (any) --> 1.000005
 - (run req) ~ namespace::autoclean (any) --> 0.22
 - (run req) ~ perl 5.010001 --> 5.010000
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (conf req) + ExtUtils::MakeMaker (any)
 - (conf req) - Module::Build::Tiny
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::CheckExtraTests
 - (dev req) - Dist::Zilla::Plugin::ContributorsFromGit
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::ModuleBuildTiny
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1407
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0309

0.0103 2015-03-19T18:46:54Z CET
 [Bug Fixes]
 - Fixed a markdown syntax error.

 [Documentation]
 - Added link to Task::Badge::Depot.

 [Meta]
 - Updated iller.yaml syntax.

0.0102 2015-02-18T14:23:27Z CET
 [Documentation]
 - Fixed links to Types::URI.

0.0101 2015-02-17T21:35:56Z CET
 [Meta]
 - Removed (unnecessary) cpanfile.

 [Requirements]
 - Listed as prereq: MooseX::AttributeShortcuts.
 - Listed as prereq: namespace::autoclean.
 - Added a dev prereq (in order to link to github from pod).

0.0100 2015-02-17T21:19:06Z CET
 - Initial release


t/corpus/lib/Badge/Depot/Plugin/Afakebadge.pm
use Badge::Depot::Standard;

class Badge::Depot::Plugin::Afakebadge using Moose with Badge::Depot {

    has username => (
        is => 'ro',
        isa => Str,
    );
    has repo => (
        is => 'ro',
        isa => Str,
    );
    has alt_text => (
        is => 'ro',
        isa => Str,
        predicate => 1,
    );
    has dont_link => (
        is => 'ro',
        isa => Bool,
        default => 0,
    );

    method BUILD {
        $self->link_url(sprintf q{https://travis-ci.org/%s/%s}, $self->username, $self->repo) if !$self->dont_link;
        $self->image_url(sprintf q{https://travis-ci.org/%s/%s.svg?branch=master}, $self->username, $self->repo);
        $self->image_alt($self->alt_text) if $self->has_alt_text;
    }
}

1;


t/corpus/lib/Badge/Depot/Plugin/Afakebadgewithoutimage.pm
use Badge::Depot::Standard;

class Badge::Depot::Plugin::Afakebadgewithoutimage using Moose with Badge::Depot {

    has username => (
        is => 'ro',
        isa => Str,
    );

    method BUILD {
        $self->link_url(sprintf q{https://travis-ci.org/%s}, $self->username);
    }
}


lib/Badge/Depot.pm
use strict;
use warnings;

package Badge::Depot;

# ABSTRACT: A framework for badges
# AUTHORITY
our $VERSION = '0.0105';

use Moose::Role;
use Types::URI qw/Uri/;
use Types::Standard qw/Str InstanceOf/;
use MooseX::AttributeShortcuts;
use namespace::autoclean;

has image_url => (
    is => 'rw',
    isa => Uri,
    init_arg => undef,
    coerce => 1,
    predicate => 1,
);
has image_alt => (
    is => 'rw',
    isa => Str,
    init_arg => undef,
    predicate => 1,
);
has link_url => (
    is => 'rw',
    isa => Uri,
    init_arg => undef,
    coerce => 1,
    predicate => 1,
);
has zilla => (
    is => 'ro',
    isa => InstanceOf['Dist::Zilla'],
    predicate => 1,
);


around qw/to_html to_markdown/ => sub {
    my $next = shift;
    my $self = shift;

    return '' if !$self->has_image_url;
    $self->$next;
};

sub to_html {
    my $self = shift;

    my $image_text = $self->has_image_alt ? sprintf ' alt="%s"', $self->image_alt : '';

    if($self->has_link_url) {
        return sprintf q{<a href="%s"><img src="%s"%s /></a>}, $self->link_url, $self->image_url, $image_text;
    }
    else {
        return sprintf q{<img src="%s"%s />}, $self->image_url, $image_text;
    }
}
sub to_markdown {
    my $self = shift;

    if($self->has_link_url) {
        return sprintf q<[![%s](%s)](%s)>, $self->image_alt // '', $self->image_url, $self->link_url;
    }
    else {
        return sprintf q<![%s](%s)>, $self->image_alt // '', $self->image_url;
    }
}

1;

__END__

=pod

=head1 SYNOPSIS

    # Define a badge class
    package Badge::Depot::Plugin::Example;

    use Moose;
    with 'Badge::Depot';

    has user => (
        is => 'ro',
        isa => 'Str',
        required => 1,
    );

    sub BUILD {
        my $self = shift;
        $self->link_url(sprintf 'https://example.com/users/%s', $self->user);
        $self->image_url(sprintf 'https://example.com/users/%s.svg', $self->user);
        $self->image_alt('Example text');
    }

    # Somewhere else
    my $badge = Badge::Depot::Plugin::Example->new(user => 'my_username');

    print $badge->to_html;
    # prints '<a href="https://example.com/users/my_username"><img src="https://example.com/users/my_username.svg" alt="Example text" /></a>'


=head1 DESCRIPTION

C<Badge::Depot> is a framework for documentation badges. Using badges in your documentation can give
end users of your distribution dynamic information without you having to update the documentation.

You only need use this distribution directly if you want to create a new badge in the C<Badge::Depot::Plugin> namespace. See L<Task::Badge::Depot> for
a list of existing badges.

For use together with L<Pod::Weaver>, see L<Pod::Weaver::Section::Badges>.

=head1 OVERVIEW

C<Badge::Depot> is a L<Moose> role that adds a few attributes and methods.


=head1 ATTRIBUTES

These attributes are expected to be set when the badge class returns from C<new>. See L<synopsis|/SYNOPSIS>.


=head2 image_url

Required. L<Uri|Types::URI>.

The url to the actual badge. The src attribute for the img tag when rendered to html.


=head2 image_alt

Optional. L<Str|Types::Standard>.

The alternative text of the badge. The alt attribute for the img tag when rendered to html. No alternative text is created if this isn't set.


=head2 link_url

Optional (but recommended). L<Uri|Types::URI>.

The url to link the badge to. The href attribute for the a tag when rendered to html. No link is created if this isn't set.


=head2 zilla

This is not a user setable attribute, but rather can be used by plugins that might need to look at distribution meta data. This is only usable when using L<Pod::Weaver::Section::Badges> together with L<Dist::Zilla>.

=head1 METHODS

These methods are used when rendering the badge, and are not useful inside badge classes.

=head2 to_html

Returns a string with the badge rendered as html.

=head2 to_markdown

Returns a string with the badge rendered as markdown.


=head1 SEE ALSO

=for :list
* L<Task::Badge::Depot> - List of available badges
* L<Pod::Weaver::Section::Badges> - Weave the badges
* L<WWW::StatusBadge> - An alternative badge framework

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A framework for badges",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Badge-Depot",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Badge::Depot",
  "PREREQ_PM" => {
    "Moose::Role" => "2.0000",
    "MooseX::AttributeShortcuts" => "0.028",
    "Types::Standard" => "1.000005",
    "Types::URI" => 0,
    "namespace::autoclean" => "0.22"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Moose" => "2.0000",
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0105",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moose" => "2.0000",
  "Moose::Role" => "2.0000",
  "MooseX::AttributeShortcuts" => "0.028",
  "Test::More" => "0.96",
  "Types::Standard" => "1.000005",
  "Types::URI" => 0,
  "namespace::autoclean" => "0.22"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
